### PR TITLE
[Operator] Add uniform, normal, resolve_neg & resolve_conj & zeros & ones & full Ops with UT & Bench

### PR DESCRIPTION
**Pull request by Bowen12992**
- Add uniform, normal, resolve_neg & resolve_conj Ops with UT & Bench.
- Here I want to reuse code with RNG funcs, but triton does not support high-order function https://github.com/triton-lang/triton/issues/4373, this PR may be simplified with high-order funcs.
- Only the performance of normal op is not as good as `Aten` with small-size input because it currently use two kernels.
- Add ones/zeros/full/full_like/ones_like/zeros_like Ops with UT & Bench.
- Add UT of K-S test but it may fail randomly

### CODE REVIEW

File: src/flag_gems/ops/normal.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+@triton.jit
+def transform_func_float_tensor(val, std, mean):
+    return val * std + mean
+
+
+@pointwise_dynamic(
+    is_tensor=[True, False, False], promotion_methods=[(0, 1, 2, "DEFAULT")]
+)
+@triton.jit
+def transform_func_float_float(val, std, mean):
```
Comment (tongxin) [inline src/flag_gems/ops/normal.py:42]: The function name is not informative.

File: src/flag_gems/ops/normal.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    return val * std + mean
+
+
+def get_std_normal(mean, std, *, generator=None):
```
Comment (tongxin) [inline src/flag_gems/ops/normal.py]: Standard normal distribution has dedicated mean and variance. It's a bit strange passing mean and std as parameters again.

File: src/flag_gems/ops/uniform.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    c0 = (philox_offset & 0xFFFFFFFF).to(tl.uint32)
+    c1 = ((philox_offset >> 32) & 0xFFFFFFFF).to(tl.uint32)
+    i4 = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
+    c0 += i4
+    _O = c0 * 0
+    r0, r1, r2, r3 = tl.philox(philox_seed, c0, c1, _O, _O)
+    r0 = uint_to_uniform_float(r0) * (to - from_) + from_
+    r1 = uint_to_uniform_float(r1) * (to - from_) + from_
+    r2 = uint_to_uniform_float(r2) * (to - from_) + from_
+    r3 = uint_to_uniform_float(r3) * (to - from_) + from_
```
Comment (tongxin) [inline src/flag_gems/ops/uniform.py:54]: We need to outline this chunk of code into a function in random_utils.

File: src/flag_gems/ops/resolve_conj.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import logging
+
+import torch
+
+from flag_gems.ops.neg import neg_func
+
+
+def resolve_conj(A: torch.Tensor):
+    logging.debug("GEMS RESOLVE_CONJ")
+    return neg_func(A.imag) if A.is_conj() else A
```
Comment (tongxin) [inline src/flag_gems/ops/resolve_conj.py]: The true branch only returns imaginary part of the complex tensor.

Comment (tongxin) [reply src/flag_gems/ops/resolve_conj.py]: Before a conjugate is materialized, it shares the original complex tensor's data. Resolving a conjugate complex probably requires you make a copy of the whole tensor in the first place then negate its imag part.

File: src/flag_gems/ops/full.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    BLOCK_SIZE: tl.constexpr,
+):
+    pid = tl.program_id(axis=0)
+    block_start = pid * BLOCK_SIZE
+    offsets = block_start + tl.arange(0, BLOCK_SIZE)
+    mask = offsets < n_elements
+    tl.store(output_ptr + offsets, fill_value, mask=mask)
+
+
+def full(size, fill_value, *, dtype=None, layout=None, device=None, pin_memory=None):
```
Comment (StrongSpoon) [inline src/flag_gems/ops/full.py:24]: are layout and pin_memory not used here?

File: src/flag_gems/ops/full.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import logging
+
+import torch
+import triton
+import triton.language as tl
+
+from flag_gems.utils.shape_utils import volume
+
+
+@triton.jit
```
Comment (StrongSpoon) [inline src/flag_gems/ops/full.py]: fill_value doesn't need specialization

File: src/flag_gems/ops/full.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+def full_kernel(
+    output_ptr,
+    n_elements,
+    fill_value,
+    BLOCK_SIZE: tl.constexpr,
```
Comment (StrongSpoon) [inline src/flag_gems/ops/full.py:15]: a tunable block_size might bring better performance

File: tests/test_special_ops.py [hunk 8]
```diff
@@ -222,3 +222,102 @@ def test_accuracy_rand_like(shape, dtype):
...
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_normal(shape, dtype):
+    loc = torch.full(size=shape, fill_value=3.0, dtype=dtype, device="cuda")
+    scale = torch.full(size=shape, fill_value=10.0, dtype=dtype, device="cuda")
+    with flag_gems.use_gems():
+        res_out = torch.distributions.normal.Normal(loc, scale).sample()
+    mean = torch.mean(res_out)
+    std = torch.std(res_out)
+    assert torch.abs(mean - 3.0) < 0.1
+    assert torch.abs(std - 10.0) < 0.1
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: I suggest a convincing method as assertion.

File: tests/test_special_ops.py [hunk 9]
```diff
@@ -222,3 +222,102 @@ def test_accuracy_rand_like(shape, dtype):
...
+    y = x.conj()
+    assert y.is_conj()
+    with flag_gems.use_gems():
+        z = y.resolve_conj()
+    assert not z.is_conj()
+
+
+@pytest.mark.parametrize("shape", POINTWISE_SHAPES)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_zeros(shape, dtype):
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: zeros/ones/full could be divided into pointwise operators

File: tests/test_special_ops.py [hunk 10]
```diff
@@ -222,3 +222,102 @@ def test_accuracy_rand_like(shape, dtype):
...
+    with flag_gems.use_gems():
+        res_out = torch.zeros(shape, dtype=dtype, device="cuda")
+    assert torch.equal(res_out, torch.zeros(shape, dtype=dtype, device="cuda"))
```
Comment (StrongSpoon) [inline tests/test_special_ops.py]: gems_assert_equal

**Review by StrongSpoon — COMMENTED** [review]
I'm working on a survey on K-S test. We should test the accuracy of distribution operators in a more convincing way.

File: tests/test_special_ops.py (hunk 10 above)
Comment (Bowen12992) [reply tests/test_special_ops.py]: ok

File: tests/test_special_ops.py (hunk 9 above)
Comment (Bowen12992) [reply tests/test_special_ops.py]: i will add  op_creation & distribution unit test files

File: tests/test_special_ops.py (hunk 8 above)
Comment (Bowen12992) [reply tests/test_special_ops.py]: Ok

File: src/flag_gems/ops/full.py (hunk 7 above)
Comment (Bowen12992) [reply src/flag_gems/ops/full.py:15]: `full`,`zeros`, `ones` is pointwsie-like operator, a tunable block_size has little effect

File: src/flag_gems/ops/full.py (hunk 6 above)
Comment (Bowen12992) [reply src/flag_gems/ops/full.py]: sure

File: src/flag_gems/ops/full.py (hunk 5 above)
Comment (Bowen12992) [reply src/flag_gems/ops/full.py:24]: I just keep the interface same with aten, Actually , only c-contiguous array is supported

**Comment by Bowen12992** [comment]
> since k-s test is already used for exponential, it could be expanded on normal and uniform as well.

sure will add that later

File: src/flag_gems/ops/normal.py (hunk 1 above)
Comment (Bowen12992) [reply src/flag_gems/ops/normal.py:42]: This name consist with aten name `tensor_tenosr`_normal, `tensor_float_normal`, `float_float_normal`

File: src/flag_gems/ops/normal.py (hunk 2 above)
Comment (Bowen12992) [reply src/flag_gems/ops/normal.py]: I will change the name

File: src/flag_gems/ops/uniform.py (hunk 3 above)
Comment (Bowen12992) [reply src/flag_gems/ops/uniform.py:54]: triton does not support high-order function, use tow function will cause performance decrease

File: src/flag_gems/ops/resolve_conj.py (hunk 4 above)
Comment (Bowen12992) [reply src/flag_gems/ops/resolve_conj.py]: will be fixed later，thanks～

File: src/flag_gems/ops/full.py (hunk 7 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/full.py:15]: vendors might care about this. but that's okay.

File: src/flag_gems/ops/normal.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    philox_seed, philox_offset = philox_cuda_seed_offset(increment)
+    with torch.cuda.device(mean.device):
+        randn_kernel[grid_fn](out, N, philox_seed, philox_offset)
+    return out
+
+
+def normal_tensor_tensor(mean, std, *, generator=None):
+    logging.debug("GEMS NORMAL_TENSOR_TENSOR")
+    out = normal_distribution(mean, std)
+    return transform_func_tensor_tensor(out, std, mean)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/normal.py:62]: normal needs to launch two kernels. suggest fusing them together.

File: src/flag_gems/ops/resolve_conj.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import logging
+
+import torch
+
+
+def resolve_conj(A: torch.Tensor):
+    logging.debug("GEMS RESOLVE_CONJ")
+    return torch.complex(A.real, A.imag.neg()) if A.is_conj() else A
```
Comment (StrongSpoon) [inline src/flag_gems/ops/resolve_conj.py:8]: will A.imag.neg fall into flag_gems.neg?

Comment (Bowen12992) [reply src/flag_gems/ops/resolve_conj.py:8]: yes

File: src/flag_gems/ops/normal.py (hunk 11 above)
Comment (Bowen12992) [reply src/flag_gems/ops/normal.py:62]: OK，will try that later

**Review by StrongSpoon — APPROVED** [review]
lgtm.
